### ISSUE TITLE
Implement return value from yield()

### ISSUE THREAD

**Issue by Spooner** [godotengine/godot#1397]
To create generator functions, we need yield(value) to be implemented. This would act like yield(), but 'value' would be passed back and returned from the resume() call.

This would allow range() to be made into an iterator, rather than creating a potentially huge array.

This would also be a step towards nice features like list/dict comprehension.

**Comment by jamie-pate** [comment]
Unpleasant workaround: 
```
func xrange(count, dict):
	var i = 0
	while i < count:
 		dict.next = i
		yield()

func do_stuff():
	var curr = {}
	var it = xrange(curr)
	while curr.value != null:
		print(curr.value)
		it.resume()
```

**Comment by busy8** [comment]
I've just started working with Godot (liking it so far) but I agree that it feels that yield() should be able to return more than the function state. Yield is like returning with the ability to resume() so the function should be able to return a value.

One possible approach I have been thinking of is augmenting [GDScript]FunctionState 
( https://docs.godotengine.org/en/3.1/classes/class_gdscriptfunctionstate.html ) 
to have a method value() which returns the argument to yield().

```
func co1:
     yield(5)

s = co1()
print(s.value())   # prints 5
```
GDscript already has a yield which takes two values and waits on a signal. However, this one argument version would just return its argument in the FunctionState. With just that change a generator could be written like:

```
func xrange(count):
	var i = 0
	while i < count:
		yield(i)

func do_stuff():
	var it = xrange(5)
	while it:
		print(it.value())
		it.resume()
```

This works quite nicely but there is a subtlety worth considering. When the coroutine function returns (instead of yielding) a value is returned not a FunctionState. In the above example null is returned (because the function is void) which we test to stop the loop in do_stuff(). If the coroutine function returns a value [eg. imagine `return false` at the end of xrange()] then that test would break. 

We also need to consider the version of yield() which listens for a signal.  I consider the reuse of the keyword "yield" slightly unfortunate here as it is more of a "wait". They could have an optional third argument that is returned as the FunctionState's value() whenever the yield() is blocked waiting for that signal. (However there is apparently a bug that causes coroutines to continue without the signal being fired when resumed using the FunctionState: https://github.com/godotengine/godot/issues/28942)

Finally (I have to check if this is already done) arguments that accompany the signal should be treated like arguments to resume(), that is, the yield() should "return" them. eg:
```
signal hp_reduced(amount)
# elsewhere in code
var amount = yield ($Player, "hp_reduced")
```
A complication here is that we, ideally, should be able to return multiple values so we could collect into an array in that case.
```
signal moved(from, to)
var args = yield($Item, "moved")
var from = args[0]
var to = args[1]
```
I think I'll take a look at the code. 
IMO, yield(value) would be, erm, valuable.

**Comment by vnen** [comment]
There's some problems on how `yield` works, so it'll replaced by `await` which is more straight-forward. I'm not sure yet whether we'll do actual generators.

If you are still interested in the idea of adding generators, please [**open a new proposal on the GIP tracker**](https://github.com/godotengine/godot-proposals/issues/new/choose) following the given issue template (after checking that it doesn't exist already). You can reference this issue and summarize the ideas in the proposal. Thanks in advance!
